pages/blog/zenwave-asyncapi-code-generator.md

```diff
@@ -0,0 +1,348 @@
+            <configuration>
+                <generatorName>spring-cloud-streams3</generatorName>
+                <configOptions>
+                    <role>provider</role>
+                    <style>imperative</style>
+                    <transactionalOutbox>mongodb</transactionalOutbox>
+                    <modelPackage>io.zenwave360.example.core.domain.events</modelPackage>
+                    <producerApiPackage>io.zenwave360.example.core.outbound.events</producerApiPackage>
+                    <consumerApiPackage>io.zenwave360.example.adapters.commands</consumerApiPackage>
+                </configOptions>
+            </configuration>
+        </execution>
+    </executions>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.github.zenwave360.zenwave-sdk.plugins</groupId>
+            <artifactId>asyncapi-spring-cloud-streams3</artifactId>
+            <version>${zenwave.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.zenwave360.zenwave-sdk.plugins</groupId>
+            <artifactId>asyncapi-jsonschema2pojo</artifactId>
+            <version>${zenwave.version}</version>
+        </dependency>
+    </dependencies>
+</plugin>
+```
+
+![ZenWave StreamCloudStreams AsyncAPI](/img/posts/zenwave/spring-cloud-streams3-generated.png)
+
+### Provider vs Client
+
+Because broker based API definitions are inherently **symmetrical** it's difficult to establish the roles of client/server. ZenWave generates code based on `provider` and `client` roles, where a `provider` "produces events" and "consumes commands". See [API-First with AsyncAPI](API-First-with-AsyncAPI) page for more details on "publish/subscribe", "producer/consumer" and "provider/client" roles.
+
+> Write your AsyncAPI definitions from the `provider` perspective and then configure the code generator to generate either a `provider` or a `client`.
+
+If you still find confusing which is a provider and a client just use this rule: In a given messaging scenario, there can be only one provider of a message, while there can be multiple clients.. If the provider is producing messages, use the `publish` section. If the provider is consuming messages, use the `subscribe` section.
+
+### Spring Cloud Streams Producer: Using generated code to produce messages
+
+On the producer side generates:
+
+- Interface `ICustomerEventsProducer` to produce typed messages that uses your domain names: `onCustomerEvent`, `CustomerEventPayload` and `CustomerEventPayloadHeaders`.
+- Producer `@Component` `CustomerEventsProducer` you can autowire in your services.
+
+**In order to produce messages all you need to do is @Autowire the generated producer as part of your code.**
+
+This producer component, because sits behind a business oriented interface, can be implemented in different flavors and integration patterns like transactional outbox (for mongodb and sql) or enterprise envelop depending on how you configure zenwave maven generator.
```

Comment:
```suggestion
Because it sits behind a business-oriented interface, this producer component can be implemented in different flavors and integration patterns like transactional outbox (for MongoDB and SQL) or enterprise envelop, depending on how you configure the Zenwave maven generator.
```